### DISCUSSION TITLE
Tidier Script Parsing

### DISCUSSION THREAD

**Discussion by trumpton** [gerbera/gerbera#2798] [Ideas]
I am looking to add scripting features to organise photographs using lightroom / darktable hierarchical subject tags.
But when looking at the way the scripts are handled / parsed, it looks to me to be rather clunky ...

1. There are three separate scripts loaded (import, playlists, metadata) each as their own script objects.
2. They all use the same underlying script class.
3. There are two class wrappers, one for import, and one for the others.
4. Scripts files are explicitly named in the config.xml.
5. There is no easy mechanism for breaking script files into more manageable lumps for development, or for selecting different layouts (at the moment, there are two layouts supported for Audio, and these are in the common.js file, which is also loaded by the playlists script even though it is not needed).

---

I propose to:

1.   Have a single script object - merging script specific support functions together - e.g. readLn will be available to all.
2.   Import media by calling js function import_media_item() rather than executing import.js
3.   Import playlists by calling js function import_playlist_item() rather than executing playlists.js
4.   Import metadata by calling js function import_metadata_item() rather than executing metadata.js
5.   Add a folder structure under the js folder, something like:

    js/import.js
    js/playlists.js
    js/metadata.js

    js/common/*.js

    js/import/default/*.js
    js/import/structured/*.js
    js/import/....

    js/playlists/default/*.js
    js/playlists/...

    js/metadata/default/*.js  
    js/playlists/...

6. Identify the folders in the config.xml, such that the 'structured' folder can be loaded after the 'default' folder, or perhaps instead of.  Perhaps there is a structured-audio, structured-video and structured-image folder, with the selection being performed in the config.xml tags.

---

Once this is completed, it will be easier to add and manage 'alternative' layout configurations to the installation.

I've also got some longer term plans to add some translation functions to the programme to provide other language support.  As some of the language is stored in the database, this goes further than simply adding a tr() function in the C++ - it must also be available in the scripts - having a single script interface will also make this easier in the future.

---

So far, I've created a branch where I've modified the code to use 'call()' instead of 'execute()' for the scripts.

https://github.com/trumpton/gerbera/tree/script_tidy

Thoughts and comments gratefully appreciated.

Steve

**Comment by KarlStraussberger** [comment]
Thank you, it's always great to have guys around who contribute in improving Gerbera.

The scripting interface is almost unchanged from its early years in MediaTomb.

I like the idea of using `call` instead of `execute` which removes implicit global variables (your changes bring them back for compatibility). This even simplifies the individual call for each media file.

I also like the idea of reading a complete directory instead of just one or two files. This allows us to use the directory as a large library. Actually, I don't see any problem if we load more functions into the runtime than necessary (that's the idea of the common file) because we only do it once at startup. That said, I would even load the `import_*_item` functions as well and have really everything available.

On the other hand we have to avoid breaking changes for users that already implemented their personal layouts or at least provide simple instructions the upgrade their scripts. 

Looking at configuration side, this means
- path to the system script directory (replacing the `common-script` path)
- name init function to call for the respective target (replacing the layout parameter and `import-script`, `playlist-script`, `metafile-script` path)
- name of the user's script directory (replacing the custom script path)

```
<scripting>
  <common-scripts>/usr/local/share/gerbera/js</common-scripts>
  <custom-scripts>/home/gerbera/js</custom-scripts>
  <media-options type="js" entry-point="import_media_item">
    <audio-layout>
      <layout-flag name="type" value="structured"/>
      <layout-flag name="skip-chars" value="" />
      <layout-flag name="album-box" value="6"/>
      <layout-flag name="artist-box" value="9"/>
      <layout-flag name="genre-box" value="6"/>
      <layout-flag name="track-box" value="6"/>
      <layout-flag name="div-char" value="-" />
    </audio-layout>
  </media-options>
  <playlist-options entry-point="import_playlist_item">
    <layout-flag name="create-link" value="yes"/>
  </playlist-options>
  <metadata-options  entry-point="import_metadata_item">
  </metadata-options>
</scripting>
```

Maybe we can even move the logic to select the media type into code, so we have
```
<audio-options type="js" entry-point="import_audio_item_structured">
...
</audio-options>

<video-options type="js" entry-point="import_video_item">
...
</video-options>

<image-options type="buildin">
...
</image-options>
```

As you see, you triggered a bunch of new ideas.

**Comment by trumpton** [comment]
I was thinking along the same sort of lines as you, and have coded up something that reads the following:

```
<scripting script-charset="UTF-8" >
      <common-script-folder>/usr/local/share/gerbera/js/common</common-script-folder>
      <playlist-script-folder>/usr/local/share/gerbera/js/playlists</playlist-script-folder>
      <metafile-script-folder>/usr/local/share/gerbera/js/metadata</metafile-script-folder>
      <virtual-layout type="js" audio-layout="search">
              <import-script>/usr/local/share/gerbera/js/import/import.js</import-script>
              <import-script-folder>/usr/local/share/gerbera/js/import</import-script-folder>
      </virtual-layout>
    </scripting>

```

For the custom.js script, I think the right thing to do is support loading of a custom.js file and/or a custom folder - that way, legacy code can be supported.

I've still got a few things I don't like - firstly, the audio-layout="blah" - I much prefer what you've suggested with the layout flags, because it is extendable.

I'm not a fan of loading all of the possible script files that may exist in all the 'option' folders, just in case someone who writes a 'structured' layout has a similarly named function as someone who writes a 'default' layout, so I'd like to see just those files associated with an option structured/default/... etc. loaded.

I also don't like the way I've had to build up the path, for example if I select _audio-layout="structured"_, the script has to load files from the _/usr/local/share/gerbera/js/import/audio-layout-structured_ folder.

Perhaps the thing to do is to include the full path to the folder to use, rather than name the subdirectory.

```
<scripting enabled="yes" type="js">

  <scripts>
    <script-folder>/usr/local/share/gerbera/js/common</script-folder>
    <script-folder>/home/gerbera/js/custom</script-folder>
    <script-file>/home/gerbera/js/custom.js</script-file>
  </scripts>

  <media-import entry-point="import_media_item">

    <scripts>
      <script-file>/usr/local/share/gerbera/js/import.js</script-file>
      <script-folder>/usr/local/share/gerbera/js/import/audio-structured</script-folder>
      <script-folder>/usr/local/share/gerbera/js/import/image-default</script-folder>
      <script-folder>/usr/local/share/gerbera/js/import/video-structured</script-folder>
      <script-folder>/usr/local/share/gerbera/js/import/trailer-default</script-folder>
    </scripts>

    <flags>
      <flag name="audio-structured-skip-chars" value=""/>
      <flag name="audio-structured-track-box" value="6"/>
      <flag name="video-default-ignoremkv" value="yes"/>
    </flags>
  
  </media-import>

  <playlist-import entry-point="import_playlist_item">

    <scripts>
      <script-folder>/usr/local/share/gerbera/js/playlists/default</script-folder>
    </scripts>

    <flags>
      <flag name="create-link" value="yes"/>
    </flags>

  </playlist-import>

  <metadata-import entry-point="import_metadata_item">

    <scripts>
      <script-folder>/usr/local/share/gerbera/js/metadata/default</script-folder>
    </scripts>

    <flags></flags>

  </metadata-import>

</scripting>

```

Whatever is adopted, I think some form of symmetry is wise, and a mechanism which is extendible (at the moment, we have audio,image,video,trailer, but at some stage in the future we may have other things.

I note that the flags section in the config file is very user specific, and when a new release is produced, it is likely that the config file will need to be re-generated, thereby losing any configuration a user has.

**Q: ** What do you think about the user being able to modify the config directly in the custom.js script rather than adding tags to the config.xml file - this way, his modifications will be safe following upgrades and regeneration of the config.xml file itself.

Another advantage could be that an import script which needs certain features can enable them in the script itself - for example, if I write a replacement addImage() json module, and need certain exiv2 tags importing, rather than identifying the tags in the xml file, they can be identified at the top of the myimportimage.js file - so the user simply needs to activate the json file, and not also do a parallel edit of the config.xml file.

So the custom.js file could contain things like:

```
    addConfigOption("scripting/media-import/flags/flag", "audio-structured-track-box", "6") ;
```

Or the myimportimage.js file could contain things like:

```
  addConfigOption("libexif/auxdata/add-data", "tag", "EXIF_TAG_MODEL") ;
  function addImage(obj)

```
Instead of config.xml containing:

```
<libexif>
    <auxdata>
        <add-data tag="EXIF_TAG_MODEL"/>
    </auxdata>
</libexif>

```
**Q:** I don't understand why the type=js gets buried down in the config tree - should it not envelope the common.js scripts too?  Is there a credible scenario where someone would use playlists.js, but still want the builtin import mechanism?

**Q:** I've elected to explicitly point to each `<script-folder>` in the above config example, as I believe it makes things simpler in the long run - i.e. the folder is the folder, and does not have to be built from a folder tag and a subdirectory attribute.  Is this approach compatible with the web config file editor strategy?

**Q:** I've (probably naiively) lumped `<script-folder>` tags together with `<script-file>` tags under a `<scripts>` tag - is this actually feasible?  I'm looking for a way of being able to add new folders to support future media types without having to re-structure the server (perhaps apart from the --create-config part.  Maybe all the sub-entries could simply be `<script>` and if it is a js file, load() is used, and if it is a folder, loadFolder() is used.

What I'll do is check in what I've got (script-tidy-multifile) so you can see what multiple files could look like - it's not representative of what is discussed above, as I created it earlier this week, however, suddenly, the common-scripts is much more manageable - I also see some functions in the common.js that aren't actually used anywhere - they could be moved to a deprecated.js, which logs a warning to the log file on the first call.

_I note that js_functions.cc has 2 instances of duk_get_global() which need to be replaced with a passed parameter.  This may also be the case in other js callback functions in other files too._
```
        duk_get_global_string(ctx, "object_script_path");
        duk_get_global_string(ctx, self->getOrigName().c_str());
```
This may make the api call to addCdsObject() change so there will be a backwards-compatibility issue here.

In the meanwhile, I've updated the `script_tidy` branch with a work-around (making orig global in import.js).

**Q:** Is nesting the objects a stupid idea - i.e. have the object passed to the import function having a copy of itself (as obj.orig).  That way, the original structure is passed around with the object, and can be used as necessary, and fields within obj can be overridden / updated as desired, and the original fields can also be accessed too (see [scripttidy-obj_orig](https://github.com/trumpton/gerbera/tree/scripttidy-obj_orig):

```
obj.id = 9
obj.refID = 9
obj.title = "Division Bell"
obj.object_script_path = "/path/to/media"
obj.upnpclass = object.item.audioItem
    obj.id = 9
    obj.refID = -1
    obj.orig.title = "01_divisionbell.mp3"
    obj.orig.object_script_path = "/path/to/media"
    obj.orig.upnpclass = object.item.audioItem
    obj.orig.orig = null
```

**Q:** Where does the obj.refID fit in here - I can see that it is set to match the obj.id in `import.js` - does it need to be?  Is there a case where it should not be set to obj.id?  Could it be pre-defined in the cc script?

**Q:** Final though / question (at least for this round) - which of the "global variable" passed in the execute call are really needed - I've got an interface that is something like this: `function import_media_item(orig, script, autoscanid, typeaudio, typeimage, typevideo)`- but I don't see typeaudio etc. being used.  In fact, I see similar strings hard coded in the js files, and I see other types which aren't passed.  Can we rationalise what is passed to the script without breaking things?  And/Or is there something that is missing that we should be adding to help future-proof?

S.

**Comment by KarlStraussberger** [comment]
I incorporated some of your ideas in #2824 

Some of the changes may even answer your questions.

I personally don't think we should split up the js files into smithereens because we load everything into the runtime anyway. Also it breaks existing configurations and force users to patch their scripts and configs.
